airflow/jobs/backfill_job.py

```diff
@@ -341,13 +341,13 @@ def _task_instances_for_dag_run(self, dag_run, session=None):
         dag_run.refresh_from_db()
         make_transient(dag_run)
 
-        # TODO(edgarRd): AIRFLOW-1464 change to batch query to improve perf
         for ti in dag_run.get_task_instances():
             # all tasks part of the backfill are scheduled to run
             if ti.state == State.NONE:
-                ti.set_state(State.SCHEDULED, session=session)
+                ti.set_state(State.SCHEDULED, session=session, commit=False)
             if ti.state != State.REMOVED:
                 tasks_to_run[ti.key] = ti
+        session.commit()
```

Comment:
I think it will be safer if you  surround it with try/except and perform rollback in except before re-raising(). Otherwise you might get some partial back-filling if it fails  somewhere between. This would make the whole back-fill atomic.

Reply:
Thanks for pointing that out.

Reply:
fixed.